File: multi-processor-programming-study/exercises/cpp/01_basics/exercise03_synchronization.cpp
/**
 * Exercise 03: Basic Synchronization
 *
 * CONCEPT: Using std::mutex and std::lock_guard for mutual exclusion
 *
 * C++20 provides std::mutex for mutual exclusion. RAII wrappers like
 * std::lock_guard and std::unique_lock ensure exception-safe locking.
 *
 * LEARNING OBJECTIVES:
 * - Use std::mutex for mutual exclusion
 * - Understand RAII and exception safety
 * - Use std::lock_guard and std::unique_lock
 *
 * TODO: Make the counter thread-safe using std::mutex
 */

#include <iostream>
#include <thread>
#include <vector>
#include <mutex>
#include <chrono>

namespace multiprocessor::basics {

/**
 * TODO: Implement a thread-safe counter using std::mutex
 */
class SafeCounter {
private:
    int count = 0;
    // TODO: Add a mutex member variable
    // HINT: std::mutex mtx;

public:
    /**
     * TODO: Make this method thread-safe using std::lock_guard
     */
    void increment() {
        // TODO: Lock the mutex using std::lock_guard
        // TODO: Increment the count
    }

    /**
     * TODO: Make this method thread-safe as well
     */
    int get_count() {
        // TODO: Should this be synchronized? Why or why not?
        return count;
    }

    /**
     * TODO: Implement a decrement method (also thread-safe)
     */
    void decrement() {
        // TODO: Implement this
    }
};

/**
 * TODO: Implement a thread-safe bank account
 *
 * This class represents a bank account that can be accessed by multiple threads.
 * It must ensure that:
 * - Balance never goes negative (if initial balance is non-negative)
 * - Deposits and withdrawals are atomic
 * - Balance queries are consistent
 */
class BankAccount {
private:
    double balance;
    // TODO: Add a mutex

public:
    explicit BankAccount(double initial_balance) : balance(initial_balance) {}

    /**
     * TODO: Implement thread-safe deposit
     * @param amount Amount to deposit (must be positive)
     */
    void deposit(double amount) {
        // TODO: Implement with synchronization
        // HINT: Validate amount is positive
    }

    /**
     * TODO: Implement thread-safe withdrawal
     * @param amount Amount to withdraw
     * @return true if withdrawal successful, false if insufficient funds
     */
    bool withdraw(double amount) {
        // TODO: Implement with synchronization
        // HINT: Check if balance >= amount before withdrawing
        return false;
    }

    /**
     * TODO: Implement thread-safe balance query
     */
    double get_balance() {
        // TODO: Implement with synchronization
        return 0.0;
    }
};

void test_safe_counter() {
    std::cout << "=== Testing SafeCounter ===\n";

    SafeCounter counter;
    const int num_threads = 10;
    const int increments_per_thread = 1000;

    std::vector<std::thread> threads;
    for (int i = 0; i < num_threads; ++i) {
        threads.emplace_back([&counter, increments_per_thread]() {
            for (int j = 0; j < increments_per_thread; ++j) {
                counter.increment();
            }
        });
    }

    for (auto& t : threads) {
        t.join();
    }

    int expected = num_threads * increments_per_thread;
    int actual = counter.get_count();

    std::cout << "Expected: " << expected << "\n";
    std::cout << "Actual: " << actual << "\n";
    std::cout << (actual == expected ? "✅ PASS\n" : "❌ FAIL\n");
}

void test_bank_account() {
    std::cout << "\n=== Testing BankAccount ===\n";

    BankAccount account(1000.0);

    // Multiple threads depositing and withdrawing
    std::vector<std::thread> threads;

    // 10 threads depositing
    for (int i = 0; i < 10; ++i) {
        threads.emplace_back([&account]() {
            for (int j = 0; j < 100; ++j) {
                account.deposit(10.0);
            }
        });
    }

    // 10 threads withdrawing
    for (int i = 0; i < 10; ++i) {
        threads.emplace_back([&account]() {
            for (int j = 0; j < 100; ++j) {
                account.withdraw(10.0);
            }
        });
    }

    for (auto& t : threads) {
        t.join();
    }

    double expected_balance = 1000.0;
    double actual_balance = account.get_balance();

    std::cout << "Expected balance: $" << expected_balance << "\n";
    std::cout << "Actual balance: $" << actual_balance << "\n";
    std::cout << (std::abs(actual_balance - expected_balance) < 0.01 ? "✅ PASS\n" : "❌ FAIL\n");
}

} // namespace multiprocessor::basics

/**
 * Demonstration and testing
 */
int main() {
    using namespace multiprocessor::basics;

    test_safe_counter();
    test_bank_account();

    return 0;
}

File: multi-processor-programming-study/exercises/java/src/main/java/com/multiprocessor/basics/Exercise03_Synchronization.java
package com.multiprocessor.basics;

/**
 * Exercise 03: Basic Synchronization
 *
 * CONCEPT: Using synchronized keyword to prevent race conditions
 *
 * Java provides the 'synchronized' keyword to ensure mutual exclusion -
 * only one thread can execute a synchronized block/method at a time.
 *
 * LEARNING OBJECTIVES:
 * - Use synchronized methods
 * - Understand mutual exclusion
 * - Fix race conditions with synchronization
 *
 * TODO: Make the counter thread-safe using synchronization
 */
public class Exercise03_Synchronization {

    /**
     * TODO: Implement a thread-safe counter using synchronized
     */
    public static class SafeCounter {
        private int count = 0;

        /**
         * TODO: Make this method thread-safe using 'synchronized' keyword
         */
        public void increment() {
            // TODO: Add synchronized and implement increment
            // HINT: Use 'synchronized' keyword in the method signature
        }

        /**
         * TODO: Make this method thread-safe as well
         */
        public int getCount() {
            // TODO: Should this be synchronized too? Why or why not?
            return count;
        }

        /**
         * TODO: Implement a decrement method (also thread-safe)
         */
        public void decrement() {
            // TODO: Implement this
        }
    }

    /**
     * TODO: Implement a thread-safe bank account
     *
     * This class represents a bank account that can be accessed by multiple threads.
     * It must ensure that:
     * - Balance never goes negative (if initial balance is non-negative)
     * - Deposits and withdrawals are atomic
     * - Balance queries are consistent
     */
    public static class BankAccount {
        private double balance;

        public BankAccount(double initialBalance) {
            this.balance = initialBalance;
        }

        /**
         * TODO: Implement thread-safe deposit
         * @param amount Amount to deposit (must be positive)
         */
        public void deposit(double amount) {
            // TODO: Implement with synchronization
            // HINT: Validate amount is positive
        }

        /**
         * TODO: Implement thread-safe withdrawal
         * @param amount Amount to withdraw
         * @return true if withdrawal successful, false if insufficient funds
         */
        public boolean withdraw(double amount) {
            // TODO: Implement with synchronization
            // HINT: Check if balance >= amount before withdrawing
            return false;
        }

        /**
         * TODO: Implement thread-safe balance query
         */
        public double getBalance() {
            // TODO: Implement with synchronization
            return 0.0;
        }
    }

    /**
     * Demonstration and testing
     */
    public static void main(String[] args) throws InterruptedException {
        System.out.println("=== Testing SafeCounter ===");
        testSafeCounter();

        System.out.println("\n=== Testing BankAccount ===");
        testBankAccount();
    }

    private static void testSafeCounter() throws InterruptedException {
        SafeCounter counter = new SafeCounter();
        int numThreads = 10;
        int incrementsPerThread = 1000;

        Thread[] threads = new Thread[numThreads];
        for (int i = 0; i < numThreads; i++) {
            threads[i] = new Thread(() -> {
                for (int j = 0; j < incrementsPerThread; j++) {
                    counter.increment();
                }
            });
            threads[i].start();
        }

        for (Thread t : threads) {
            t.join();
        }

        int expected = numThreads * incrementsPerThread;
        int actual = counter.getCount();

        System.out.println("Expected: " + expected);
        System.out.println("Actual: " + actual);
        System.out.println(actual == expected ? "✅ PASS" : "❌ FAIL");
    }

    private static void testBankAccount() throws InterruptedException {
        BankAccount account = new BankAccount(1000.0);

        // Multiple threads depositing and withdrawing
        Thread[] threads = new Thread[20];
        for (int i = 0; i < 10; i++) {
            threads[i] = new Thread(() -> {
                for (int j = 0; j < 100; j++) {
                    account.deposit(10.0);
                }
            });
        }
        for (int i = 10; i < 20; i++) {
            threads[i] = new Thread(() -> {
                for (int j = 0; j < 100; j++) {
                    account.withdraw(10.0);
                }
            });
        }

        for (Thread t : threads) {
            t.start();
        }
        for (Thread t : threads) {
            t.join();
        }

        double expectedBalance = 1000.0; // 10 threads * 100 * $10 - 10 threads * 100 * $10
        double actualBalance = account.getBalance();

        System.out.println("Expected balance: $" + expectedBalance);
        System.out.println("Actual balance: $" + actualBalance);
        System.out.println(Math.abs(actualBalance - expectedBalance) < 0.01 ? "✅ PASS" : "❌ FAIL");
    }
}
